File: java/src/main/java/com/newnancity/newnanmanager/services/PlayerService.java
package com.newnancity.newnanmanager.services;

import com.newnancity.newnanmanager.SessionContext;
import com.newnancity.newnanmanager.exceptions.NewNanManagerException;
import com.newnancity.newnanmanager.generated.api.PlayerServiceApi;
import com.newnancity.newnanmanager.generated.model.BanPlayerRequest;
import com.newnancity.newnanmanager.generated.model.CreatePlayerRequest;
import com.newnancity.newnanmanager.generated.model.ListPlayersResponse;
import com.newnancity.newnanmanager.generated.model.Player;
import com.newnancity.newnanmanager.generated.model.UpdatePlayerRequest;
import com.newnancity.newnanmanager.generated.model.ValidateRequest;
import com.newnancity.newnanmanager.generated.model.ValidateResponse;

/** Player management service using generated request and model types. */
public final class PlayerService extends AbstractService {
    private final PlayerServiceApi api;

    public PlayerService(PlayerServiceApi api) {
        this.api = api;
    }

    public Player createPlayer(CreatePlayerRequest request) throws NewNanManagerException {
        return execute(() -> api.createPlayer(request).execute());
    }

    public Player getPlayer(int id) throws NewNanManagerException {
        return execute(() -> api.getPlayer(id).execute());
    }

    public Player updatePlayer(int id, UpdatePlayerRequest request) throws NewNanManagerException {
        return execute(() -> api.updatePlayer(id, request).execute());
    }

    public void deletePlayer(int id) throws NewNanManagerException {
        executeVoid(() -> api.deletePlayer(id).execute());
    }

    public ListPlayersResponse listPlayers() throws NewNanManagerException {
        return listPlayers(null, null, null, null, null, null, null, null, null);
    }

    public ListPlayersResponse listPlayers(
            Integer page,
            Integer pageSize,
            String search,
            Integer townId,
            Long banMode,
            String name,
            String qq,
            String qqguild,
            String discord) throws NewNanManagerException {
        return execute(() -> api.listPlayers()
                .page(page)
                .pageSize(pageSize)
                .search(search)
                .townId(townId)
                .banMode(banMode)
                .name(name)
                .qq(qq)
                .qqguild(qqguild)
                .discord(discord)
                .execute());
    }

    public void banPlayer(int playerId, BanPlayerRequest request) throws NewNanManagerException {
        executeVoid(() -> api.banPlayer(playerId, request).execute());
    }

    public void unbanPlayer(int playerId) throws NewNanManagerException {
        executeVoid(() -> api.unbanPlayer(playerId).execute());
    }

    public ValidateResponse validate(
            String sessionId,
            long sessionEpoch,
            ValidateRequest request) throws NewNanManagerException {
        SessionContext session = new SessionContext(sessionId, sessionEpoch);
        return execute(() -> api.validate(session.getId(), session.getEpoch(), request).execute());
    }

    public ValidateResponse validate(
            ValidateRequest request,
            SessionContext session) throws NewNanManagerException {
        return validate(session.getId(), session.getEpoch(), request);
    }

    public PlayerServiceApi raw() {
        return api;
    }
}

File: kotlin/src/main/kotlin/com/nanmanager/bukkit/services/PlayerService.kt
package com.nanmanager.bukkit.services

import com.nanmanager.bukkit.http.HttpClient
import com.nanmanager.bukkit.SessionContext
import com.nanmanager.bukkit.models.*

/** 玩家管理服务；传输和模型解析由 OpenAPI 生成客户端负责。 */
class PlayerService(httpClient: HttpClient) : BaseService(httpClient) {
    private val api get() = httpClient.generatedApis.players

    fun createPlayer(request: CreatePlayerRequest): Player {
        val body = toGenerated<com.newnanmanager.generated.models.CreatePlayerRequest>(request)
        return generated<Player, com.newnanmanager.generated.models.Player> {
            api.createPlayerWithHttpInfo(body)
        }
    }

    fun getPlayer(request: GetPlayerRequest): Player {
        return generated<Player, com.newnanmanager.generated.models.Player> {
            api.getPlayerWithHttpInfo(request.id)
        }
    }

    fun updatePlayer(request: UpdatePlayerRequest): Player {
        val body = toGenerated<com.newnanmanager.generated.models.UpdatePlayerRequest>(request)
        return generated<Player, com.newnanmanager.generated.models.Player> {
            api.updatePlayerWithHttpInfo(request.id, body)
        }
    }

    fun deletePlayer(request: DeletePlayerRequest) {
        generatedVoid { api.deletePlayerWithHttpInfo(request.id) }
    }

    fun listPlayers(request: ListPlayersRequest): ListPlayersResponse {
        val banMode = request.banMode?.let { mode ->
            com.newnanmanager.generated.apis.PlayerServiceApi.BanModeListPlayers.values()
                .first { it.value.toInt() == mode.value }
        }
        return generated<ListPlayersResponse, com.newnanmanager.generated.models.ListPlayersResponse> {
            api.listPlayersWithHttpInfo(
                page = request.page,
                pageSize = request.pageSize,
                search = request.search,
                townId = request.townId,
                banMode = banMode,
                name = request.name,
                qq = request.qq,
                qqguild = request.qqguild,
                discord = request.discord
            )
        }
    }

    fun banPlayer(request: BanPlayerRequest) {
        val body = toGenerated<com.newnanmanager.generated.models.BanPlayerRequest>(request)
        generatedVoid { api.banPlayerWithHttpInfo(request.playerId, body) }
    }

    fun unbanPlayer(request: UnbanPlayerRequest) {
        generatedVoid { api.unbanPlayerWithHttpInfo(request.playerId) }
    }

    fun validate(request: ValidateRequest): ValidateResponse {
        return post("/api/v1/players/validate", request)
    }

    fun validate(request: ValidateRequest, session: SessionContext): ValidateResponse {
        val body = toGenerated<com.newnanmanager.generated.models.ValidateRequest>(request)
        return generated<ValidateResponse, com.newnanmanager.generated.models.ValidateResponse> {
            api.validateWithHttpInfo(session.id, session.epoch, body)
        }
    }
}
